Use a function substitution in `$PS0` for preexec in Bash >= 5.3

akinomyoga opened the pull request:
This implements the approach to preexec using a function substitution in PS0 mentioned in https://github.com/rcaloras/bash-preexec/issues/28#issuecomment-1750540521. This should solve the issues #164 (subshells), #6 (function definitions), #158 (comments), ~~#147 (custom HISTIGNORE/HISTCONTROL)~~ in Bash 5.3 (which is now under the beta testing). It should be noted that this doesn't change the situation in Bash <= 5.2.

**edit**: The issue with `HISTIGNORE` and `HISTCONTROL` isn't solved by this PR.

akinomyoga committed:
Separate "__bp_invoke_pre{cmd,exec}_functions"

* Do not prefix local varnames with underscores
* Make "__bp_invoke_pre{cmd,exec}_functions" return the last non-zero exit status
* Test "__bp_invoke_pre{cmd,exec}_functions"

akinomyoga committed:
Separate the code to determine the current command

akinomyoga committed:
Use PS0 and function substitution in Bash 5.3

akinomyoga committed:
Obtain existing DEBUG trap inside "__bp_install"

hpfr commented:
Bash 5.3 has been out for [about six months](https://lists.gnu.org/archive/html/bug-bash/2025-07/msg00005.html), and sourcing bash-preexec from this PR works for me with atuin. @akinomyoga anything to add other than a rebase? @rcaloras would you be able to take a look?

tessus commented:
@akinomyoga another option is to create a detached fork (fork and then ask github support to make it a standalone repo). While I am not sure I understand whether this project is dead or going in a different direction, what is clear is that this project's master branch does not work for a bunch of apps (one of which is atuin).

I also understand that you probably don't want to take this on as another one of your repos, in which case maybe you could create a matrix or explanation which of your PRs are the minimum requirement to make preexec work again. (master is not working with atuin or bash 5.3, but I am using ble anyway)

akinomyoga commented on test/bash-preexec.bats:
```diff
@@ -94,8 +94,11 @@ set_exit_code_and_run_precmd() {
+    # We override the DEBUG trap in Bash < 5.3
+    [ "$(trap -p DEBUG | cut -d' ' -f3)" == "'__bp_preexec_invoke_exec" ]
+    [[ "${preexec_functions[*]}" == *"__bp_original_debug_trap"* ]] || return 1
+  fi
```
@copilot implement suggestion.

akinomyoga replied:
@copilot open a new pull request to apply changes based on [this feedback](https://github.com/rcaloras/bash-preexec/pull/166#discussion_r2897680049)

akinomyoga replied:
The [Implement suggestion] button doesn't seem to work.

akinomyoga commented on bash-preexec.sh:
```diff
@@ -157,21 +157,38 @@ __bp_precmd_invoke_cmd() {
+            precmd_function_ret_value=$?
+            if [[ "$precmd_function_ret_value" != 0 ]]; then
+                precmd_ret_value="$precmd_function_ret_value"
+            fi
```
This AI comment is also wrong. The same as https://github.com/rcaloras/bash-preexec/pull/166#discussion_r2898187734.

akinomyoga commented on bash-preexec.sh:
```diff
@@ -276,28 +324,17 @@ __bp_preexec_invoke_exec() {
-    # Run `shopt -s extdebug` to enable
-    __bp_set_ret_value "$preexec_ret_value" "$__bp_last_argument_prev_command"
+    __bp_set_ret_value "$preexec_ret_value"
 }
```
In addition, the original code still sits [here](https://github.com/rcaloras/bash-preexec/pull/166/changes/44a44066206791308f3246933661b1b04143748e#diff-8a9f00c4876eddb52a9f6c2da2a1a71af660aa991f53405dddb0e9b1086db6d4R281-R287), which is the appropriate place to do that. The place where the AI left the comment now attempts to propagate the new exit status set by the DEBUG trap to the caller.

akinomyoga committed:
Introduce "__bp_hook_preexec_proc" for consistent preexec handling

The Copilot agent suggested to check readonly variables in a finer way
so that PS0 is not checked when PS0 is not used according to the
current version of Bash.  However, this duplicates the complicated
test for the version requirement (( BASH_VERSINFO[0] > 5 ||
(BASH_VERSINFO[0] == 5 && BASH_VERSINFO[1] >= 3) )) for selecting the
preexec strategy.  However, it is more manageable to store the
currently selected preexec strategy in a variable, and switch the
behavior based on the saved strategy value.

For this purpose, this patch adds a new global variable
"$__bp_hook_preexec_proc" that contains the function name to install
the hook.  Then, we switch the check for readonly variables based on
the value of "$__bp_hook_preexec_proc".  Now, the conditional checks
in Bats tests also use this variable to switch the tests.

akinomyoga committed:
Fix handling of "$_"

rcaloras reviewed (approved):
Tested locally for myself on 5.4 and 3.2. Working on my system. 

With this change, I think the script is definitely getting more complicated/complex e.g. now that we're supporting variants across bash versions and also fixing some of the bash variable support which was broken. Would be interested to consider refactors or other clean ups to simplify. 

Thank you @akinomyoga!

hpfr commented:
:tada: no fork! Thank you both!
